File: app-mvp/src/commonMain/kotlin/name/antonsmirnov/notes/presenter/thread/ThreadManager.kt
package name.antonsmirnov.notes.presenter.thread

/**
 * Schedules long-running tasks for execution
 */
interface ThreadManager {
    fun run(block: () -> Unit)
    fun waitForFinished()
}

/**
 * Blocking implementation of ThreadManager
 * (blocks current thread)
 */
class BlockingThreadManager : ThreadManager {

    override fun run(block: () -> Unit) = block()

    override fun waitForFinished() {
        // nothing
    }
}

File: app-ios/app-ios/ThreadManager.swift
//
//  ThreadManager.swift
//  app-ios
//
//  Created by Anton Smirnov on 19/07/2019.
//  Copyright © 2019 Anton Smirnov. All rights reserved.
//

import Foundation
import app_mvp

// Executes in current thread
class BlockingThreadManager : ThreadManager {
    
    private var semaphore: DispatchSemaphore?
    
    func run(block: @escaping () -> Void) {
        block()
    }
    
    func waitForFinished() {
        // nothing
    }
}

// Executes in background thread
class BackgroundThreadManager : ThreadManager {
    
    var semaphore: DispatchSemaphore?
    
    func run(block: @escaping () -> Void) {
        semaphore = DispatchSemaphore(value: 0)
        DispatchQueue.global().async {
            block()
        }
    }
    
    func waitForFinished() {
        _ = semaphore!.wait(timeout: .distantFuture)
    }
}



